### ISSUE TITLE
Cannot set OID spec field with SetOf Any

### ISSUE THREAD

**Issue by aalba6675** [wbond/asn1crypto#30]
I may be misunderstading oid spec but...why doesn't the following code work when the field is SetOf Anys but determined by OID spec to be SetOf BMPString.

(Or how do you actually create Attribute objects by setting fields directly instead of load'ing from existing data?)

Reproducer
```python
# from pkcs12.py
class Attribute(Sequence):
    _fields = [
        ('type', AttributeType),
        ('values', SetOf, {'spec': Any}),
    ]

    _oid_pair = ('type', 'values')
    _oid_specs = {
        'friendly_name': SetOfBMPString,
        'local_key_id': SetOfOctetString,
        'microsoft_csp_name': SetOfBMPString,
    }

class Attributes(SetOf):
    _child_spec = Attribute

>>>from asn1crypto import pkcs12, core
>>>x = pkcs12.Attribute()
>>>x['type'] = 'friendly_name'
>>>## I think x['values'] should now be interpreted as SetOfBMPString, no?
>>>x['values'] = [ 'hello', 'world']
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/venv3/lib/python3.5/site-packages/asn1crypto/core.py", line 2982, in __setitem__
    new_value = self._make_value(field_name, field_spec, value_spec, field_params, value)
  File "/venv3/lib/python3.5/site-packages/asn1crypto/core.py", line 3237, in _make_value
    wrapper = field_spec(value=new_value.dump(), **field_params)
  File "/venv3/lib/python3.5/site-packages/asn1crypto/core.py", line 3582, in __init__
    raise e
  File "/venv3/lib/python3.5/site-packages/asn1crypto/core.py", line 3573, in __init__
    self.__setitem__(index, child)
  File "/v6/venv3/lib/python3.5/site-packages/asn1crypto/core.py", line 3733, in __setitem__
    new_value = self._make_value(value)
  File "/venv3/lib/python3.5/site-packages/asn1crypto/core.py", line 3658, in _make_value
    type_name(self)
ValueError: Can not set a native python value to asn1crypto.core.SetOf where the _child_spec is Any – value must be an instance of Asn1Value
    while constructing asn1crypto.core.SetOf
```

**Comment by aalba6675** [comment]
If I load data directly, I also cannot get to recognise 'values' field as SetOfBMPString
```python
The following byte string should be an Attribute
>>>data = b'0%\x06\t*\x86H\x86\xf7\r\x01\t\x141\x18\x1e\n\x00h\x00e\x00l\x00l\x00o\x1e\n\x00w\x00o\x00r\x00l\x00d')
>>>x = pkcs12.Attribute.load (b'0%\x06\t*\x86H\x86\xf7\r\x01\t\x141\x18\x1e\n\x00h\x00e\x00l\x00l\x00o\x1e\n\x00w\x00o\x00r\x00l\x00d')
>>>x['type'].native
'friendly_name'
>>> ## looks good so far...
>>>x['values']
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/btrfs/richard.tmp/Projects/v6/venv3/lib/python3.5/site-packages/asn1crypto/core.py", line 2943, in __getitem__
    return self._lazy_child(key)
  File "/btrfs/richard.tmp/Projects/v6/venv3/lib/python3.5/site-packages/asn1crypto/core.py", line 2890, in _lazy_child
    child = self.children[index] = _build(*child)
  File "/btrfs/richard.tmp/Projects/v6/venv3/lib/python3.5/site-packages/asn1crypto/core.py", line 4811, in _build
    value.parse(nested_spec)
AttributeError: 'SetOf' object has no attribute 'parse'
```

**Comment by aalba6675** [comment]
More weirdness:

```python
## the following class works note ObjectIdentifier
class BMPAttr(P12.Attribute):
   _fields = [
        ('type', core.ObjectIdentifier),
        ('values', P12.SetOfBMPString),
    ]

>>> q = BMPAttr.load(data)
>>> q.debug()
  __main__.BMPAttr Object #139779137814256
      constructed universal tag 16
    Data: 0x06092a864886f70d01091431181e0a00680065006c006c006f1e0a0077006f0072006c0064
      Field "type"
        asn1crypto.pkcs12.AttributeType Object #139779137814144
          Header: 0x0609
            primitive universal tag 6
          Data: 0x2a864886f70d010914
            Native: friendly_name
      Field "values"
        asn1crypto.pkcs12.SetOfBMPString Object #139779137814480
          Header: 0x3118
            constructed universal tag 17
          Data: 0x1e0a00680065006c006c006f1e0a0077006f0072006c0064
          asn1crypto.core.BMPString Object #139779137813808
            Header: 0x1e0a
              primitive universal tag 30
            Data: 0x00680065006c006c006f
              Native: hello
          asn1crypto.core.BMPString Object #139779137813528
            Header: 0x1e0a
              primitive universal tag 30
            Data: 0x0077006f0072006c0064
              Native: world

## and following class fails note AttributeType instead of ObjectIdentifier
>>>class BMP1Attr(P12.Attribute):
   _fields = [
        ('type', pkcs12.AttributeType),
        ('values', pkcs12.SetOfBMPString),
    ]

>>> r = BMP1Attr.load(data)
>>> r
<__main__.BMP1Attr 139779137335592 b'0%\x06\t*\x86H\x86\xf7\r\x01\t\x141\x18\x1e\n\x00h\x00e\x00l\x00l\x00o\x1e\n\x00w\x00o\x00r\x00l\x00d'>
>>> r.debug()
  __main__.BMP1Attr Object #139779137335592
    Header: 0x3025
      constructed universal tag 16
    Data: 0x06092a864886f70d01091431181e0a00680065006c006c006f1e0a0077006f0072006c0064
      Field "type"
        asn1crypto.pkcs12.AttributeType Object #139779137335536
          Header: 0x0609
            primitive universal tag 6
          Data: 0x2a864886f70d010914
            Native: friendly_name
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/venv3/lib/python3.5/site-packages/asn1crypto/core.py", line 3494, in debug
    child = self._lazy_child(self._field_map[field_name])
  File "/venv3/lib/python3.5/site-packages/asn1crypto/core.py", line 2890, in _lazy_child
    child = self.children[index] = _build(*child)
  File "/venv3/lib/python3.5/site-packages/asn1crypto/core.py", line 4811, in _build
    value.parse(nested_spec)
AttributeError: 'SetOfBMPString' object has no attribute 'parse'
```

**Comment by wbond** [comment]
Thanks for the bug report and detailed info.

The issue should be resolved by 7b1cd4bd5a416a66418b657ab4d2709e754a78e6.

The problem was related to how `_oid_specs` works. If the spec for a field is not `None` or `core.Any`, then the result from `_oid_specs` is taken as the value spec, that is a spec for an ASN.1 value that is encoded inside of the origin spec.

This is used in a number of places, such as how public key info embeds RSA key info inside of a `core.BitString`.

It appears that in the refactoring during the development of the library, I missed updating the spec for `values` to `None` and creating a callback to determine the correct spec – either a known `SetOf`, or just `SetOfAny`. This logic was in place properly for the `cms` submodule.

Thanks to your bug report, I have added a test to cover this issue.
